[MGDOBR-1252] - Update Control Plane with ManagedProcessor status

vaibhavjainwiz opened the pull request:
https://issues.redhat.com/browse/MGDOBR-1252

## PR owner checklist

Please make sure that your PR meets the following requirements:

- [ ] Your code is properly formatted according to [this configuration](https://github.com/kiegroup/kogito-runtimes/tree/main/kogito-build/kogito-ide-config).  
  *Run `mvn clean verify -DskipTests` so that imports are correctly set.*
- [ ] Your commit messages are clear and reference the JIRA issue e.g: "[MGDOBR-XXX] - $clear_explanation_of_what_you_did"
- [ ] The layers in the `kustomize` folder have been updated accordingly.
- [ ] All new functionality is tested
- [ ] Pull Request title is properly formatted: `MGDOBR-XYZ Subject`
- [ ] Pull Request contains link to the JIRA issue
- [ ] Pull Request contains link to any dependent or related Pull Request
- [ ] Pull Request contains description of the issue
- [ ] Pull Request does not include fixes for issues other than the main ticket

## PR reviewer(s) checklist

- [x] Check if code and Github action workflows have been modified and if the modification is safe
- [x] If the modification is safe, add the `safe to test` label

<details>
<summary>
How to trigger pipelines and use the bots:
</summary>

- <b>Run the end to end pipeline</b>  
  Annotate the pull request with the label: `safe to test`. If you want to run the pipeline again, remove and add it again.

- <b>Rebase the pull request</b>  
  Comment with: `/rebase`.
- <b>Deploy BOT</b>
  Comment with `/deploy <dev,stable> when the PR has been merged to deploy to a target environment.

</details>

danielezonca commented on shard-operator-parent/shard-operator-v1/src/test/resources/application.properties:
```diff
@@ -18,3 +18,4 @@ event-bridge.executor.poll-interval.milliseconds=250
 event-bridge.ingress.poll-interval.milliseconds=250
 event-bridge.executor.deployment.timeout-seconds=60
 event-bridge.ingress.deployment.timeout-seconds=60
+quarkus.arc.test.disable-application-lifecycle-observers=true
```
Why this change? It looks like you are changing some default behaviour of quarkus

vaibhavjainwiz replied:
yes, with this property application would not listen for Startup/Shutdown Event in testcases. 
Currently we are having additional handling to prevent calling StartupEvent in testcases but with this change we don't need to add `@UnlessBuildProfile(value = "test")` in source classes.

danielezonca replied:
Let's continue here https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/pull/1453#discussion_r1060397988

danielezonca commented on infra-parent/infra-v2/src/main/java/com/redhat/service/smartevents/infra/v2/api/models/dto/BridgeStatusDTO.java:
```diff
@@ -16,12 +16,12 @@ public class BridgeStatusDTO {
 
     @JsonProperty("conditions")
-    private Set<ConditionDTO> conditions;
+    private List<ConditionDTO> conditions;
```
Why this change?
Do we expect duplicates conditions?

vaibhavjainwiz replied:
no, we don't have dublicate in conditions. I change it from `Set` to `List` to keep it consistent with `ProcessorStatusDTO` implementation.

danielezonca commented on shard-operator-parent/shard-operator-core/src/main/java/com/redhat/service/smartevents/shard/operator/core/providers/IstioGatewayProviderImpl.java:
```diff
@@ -17,11 +17,9 @@
 import io.quarkus.runtime.StartupEvent;
 
 @ApplicationScoped
-@UnlessBuildProfile(value = "test") // For tests we provide a mocked bean and this should not start at all.
```
Is it related with the other change in the properties? I don't understand the need for this change

vaibhavjainwiz replied:
`IstioGatewayProviderImpl.java` class call `setup(@Observes StartupEvent event)` on startup but in testcases we don't want this method to run. We usally mock its behaviour in test. There was no easy way to mock StartupEvent earlier due to which `@UnlessBuildProfile(value = "test")` annotation was added so that bean for this class would not created in `test` profile and MockedBean would be injected.

As I did some research I found that if we add `quarkus.arc.test.disable-application-lifecycle-observers=true` property then testcases would not register for StartupEvent/ShutdownEvent which provide more clearner way because we don't need to add any special handling in code to run the testcase.

If I don't use this new property new I need to do similiar handling in `IstioStartupService` because it also contain `setup(@Observes StartupEvent event)`.

danielezonca replied:
Ok but that property, as far as I understand, is disabling similar events for all the classes and not only for `IstioGatewayProviderImpl`. I would prefer to stay with the existing impl so that we are sure to impact only that specific class and moreover we avoid to have to remember to add that property ~everywhere (in this PR you do the change in at least three places so it is a bit more error prone)

manstis replied:
I have no strong opinion; so happy to follow @danielezonca 

Be aware that use of `quarkus.arc.test.disable-application-lifecycle-observers=true` may not work as expected if we/anyone ever writes tests that use `shard-operator-core` as a dependency vs part of a multi-module project. See https://github.com/quarkusio/quarkus/issues/21487.

vaibhavjainwiz replied:
done.. I had reverted my changes regarding this property and switched back to our old approach.

danielezonca commented on shard-operator-parent/shard-operator-v2/src/main/java/com/redhat/service/smartevents/shard/operator/v2/ManagedProcessorSyncServiceImpl.java:
```diff
@@ -0,0 +1,96 @@
+                .invoke(this::processDelta)
+                .subscribe().with(
+                        success -> LOGGER.debug("Successfully processed ManagedProcessor deltas"),
+                        failed -> LOGGER.debug("Failed to process ManagedProcessor deltas"));
```
Use `error`

vaibhavjainwiz replied:
done

r00ta reviewed (commented):
You might want to take a look at https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/pull/1451 (where I refactored a bit the DTOs - I've moved to a `ResourceStatusDTO`, removing the `BridgeStatusDTO` and the `ProcessorStatusDTO`)

vaibhavjainwiz commented:
> You might want to take a look at #1451 (where I refactored a bit the DTOs - I've moved to a `ResourceStatusDTO`, removing the `BridgeStatusDTO` and the `ProcessorStatusDTO`)

got it.. please merge your PR first. I will merge my PR over your's.

manstis commented:
FYI @vaibhavjainwiz @r00ta I also have this PR https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/pull/1445

This makes changes to `ConditionDTO` and its conversion but should not affect anything.

@r00ta Let's get your PR merged first then I will update my "metrics" PR to cover `BridgeService` too.

manstis commented on shard-operator-parent/shard-operator-v2/src/main/java/com/redhat/service/smartevents/shard/operator/v2/ManagedProcessorServiceImpl.java:
```diff
@@ -16,4 +19,10 @@ public void createManagedProcessor(ProcessorDTO processorDTO) {
+    @Override
+    public List<ManagedProcessor> fetchAllManagedProcessors() {
+        // @TODO
+        return null;
```
Unless this is to be completed before this PR is merged returning an empty collection may be preferential.

robpblake replied:
@vaibhavjainwiz This implementation is required, so assuming there is more to come?

vaibhavjainwiz replied:
done

vaibhavjainwiz replied:
provided its concreat implementation.

manstis commented on shard-operator-parent/shard-operator-v2/src/main/java/com/redhat/service/smartevents/shard/operator/v2/ManagedProcessorSyncServiceImpl.java:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    private void processDelta(List<ProcessorDTO> processorDTOList) {
+        for (ProcessorDTO processorDTO : processorDTOList) {
```
Something @r00ta pointed out to me on one of my PRs that may be desirable here too.

If processing of any of the items in the `List` results in an exception being thrown the remainder of the `List` will be skipped. It might be better defensive programming to catch and log unexpected exceptions within the loop to ensure other items can be processed? The same can be said for other places where you iterate loops of resources.

vaibhavjainwiz replied:
this is the not the best usecase to implement defensive programming.

manstis commented on shard-operator-parent/shard-operator-v2/src/main/java/com/redhat/service/smartevents/shard/operator/v2/ManagedProcessorSyncServiceImpl.java:
```diff
@@ -0,0 +1,96 @@
+                .stream().collect(Collectors.toMap(m -> m.getSpec().getId(), m -> m));
+
+        List<ProcessorStatusDTO> processorStatusDTOS = new ArrayList<>(processorDTOList.size());
+        for (ProcessorDTO processorDTO : processorDTOList) {
```
Defensive programming?

robpblake replied:
@manstis Can you give an example of what you'd like to see here  in terms of being defensive?

manstis replied:
@robpblake Sure, I was referring to my comment [above](https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/pull/1453#discussion_r1060439723).
```
processorDTOList.forEach(this::handleProcessor);
...
void handleProcessor(ProcessorDTO processorDTO) {
  try {
    //Do whatever that could throw an exception, even if unexpected...
  } catch(Exception e) {
    LOGGER.info("Blah blah...");
  }
}
```
Should processing of one `ProcessorDTO` fail for whatever reason the others are still given a chance.

vaibhavjainwiz replied:
please refer to my comment above.

manstis commented on shard-operator-parent/shard-operator-v2/src/main/java/com/redhat/service/smartevents/shard/operator/v2/resources/ManagedProcessorStatus.java:
```diff
@@ -1,36 +1,24 @@
     @Override
-    public int hashCode() {
-        return Objects.hash(conditions);
+    public ManagedResourceStatus inferManagedResourceStatus() {
```
Would it be better to refactor `shard-core` and `shard-v1` to remove this from `shard-v2`?

vaibhavjainwiz replied:
done

manstis commented on shard-operator-parent/shard-operator-v2/src/test/java/com/redhat/service/smartevents/shard/operator/v2/ManagerSyncServiceTest.java:
```diff
@@ -7,27 +7,21 @@
 
     @InjectMock
-    ManagedProcessorSyncService managedProcessorSyncService;
+    ManagedProcessorSyncServiceImpl managedProcessorSyncService;
```
Do you still need to inject the implementation now that you've refactored to an interface/implementation?

vaibhavjainwiz replied:
sorry.. my mistake

vaibhavjainwiz replied:
fixed now.

robpblake commented on shard-operator-parent/shard-operator-v2/src/main/java/com/redhat/service/smartevents/shard/operator/v2/ManagedProcessorSyncServiceImpl.java:
```diff
@@ -0,0 +1,96 @@
+        }
+    }
+
+    private List<ProcessorStatusDTO> transformToProcessorStatus(List<ProcessorDTO> processorDTOList) {
```
@vaibhavjainwiz I think this method is called for every `ProcessorDTO` we receive from the Control Plane? If yes, then we need to reconsider the placement of the `managedProcessorService.fetchAllManagedProcessors()` call as it will being called `N` times, where `N` is the number of `ProcessorDTO` received.

vaibhavjainwiz replied:
no, it just called once with `List<ProcessorDTO>`.

robpblake replied:
Thanks for clarifying @vaibhavjainwiz 👍

robpblake commented on shard-operator-parent/shard-operator-v2/src/main/java/com/redhat/service/smartevents/shard/operator/v2/ManagerClientImpl.java:
```diff
@@ -72,6 +73,15 @@ public Uni<List<ProcessorDTO>> fetchProcessorsForDataPlane() {
+    @Override
+    public Uni<HttpResponse<Buffer>> notifyProcessorStatus(List<ProcessorStatusDTO> processorStatusDTOs) {
+        LOGGER.debug("Notifying manager about the new status of the processors");
+        return getAuthenticatedRequest(webClientManager.put(V2APIConstants.V2_SHARD_API_BASE_PATH + "processors"), request -> request.sendJson(processorStatusDTOs))
```
@vaibhavjainwiz Lets add another constant to the `V2APIConstants` class e.g `V2APIConstants.V2_SHARD_API_PROCESSORS_PATH`  or something like that.

vaibhavjainwiz replied:
done

robpblake reviewed (changes requested):
Thanks @vaibhavjainwiz .

 It looks like there is missing implementation still required and I think we need to take a look at the placement of the call to read the list of `ManagedProcessor` CRs from K8S as we're doing it in a loop, when we can do it once at the start of the loop.

cheers,

Rob

vaibhavjainwiz commented:
Hi @robpblake 
I left it with @TODO by assuming its implementation would be provided along with the  task in while ManagedProcessorController would be implemented.

Now I had add just added the implementation of `fetchAllManagedProcessors`

manstis commented on shard-operator-parent/shard-operator-v2/src/main/java/com/redhat/service/smartevents/shard/operator/v2/resources/CustomResourceStatus.java:
```diff
@@ -0,0 +1,75 @@
+/**
+ * Common interface for Kubernetes Custom Resource status
+ */
+public abstract class CustomResourceStatus extends ObservedGenerationAwareStatus {
```
You could have kept this in `-core` and extended it differently for `-v1` and `-v2`.

vaibhavjainwiz replied:
I moved back `CustomeResourceStatus` to `-core`

manstis reviewed (commented):
Please advise what (if anything) you intend to do about the defensive programming comment.

I also wonder whether it's better to keep `CustomResourceStatus` in `-core` and extend it differently for `-v1` and `-v2` vs duplicating a reasonable size block of code.

robpblake commented on shard-operator-parent/shard-operator-v2/src/main/java/com/redhat/service/smartevents/shard/operator/v2/ManagedProcessorServiceImpl.java:
```diff
@@ -16,4 +25,9 @@ public void createManagedProcessor(ProcessorDTO processorDTO) {
     }
+
+    @Override
+    public List<ManagedProcessor> fetchAllManagedProcessors() {
```
@vaibhavjainwiz Could you add a unit test for this please. Thanks.

vaibhavjainwiz replied:
I had the testcase for this class.

robpblake reviewed (approved):
Minor comments from me now @vaibhavjainwiz . Once addressed, I'm happy to +1.

cheers,

Rob

r00ta commented:
Added `do not merge` label: I will remove it once https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/pull/1451 gets merged

vaibhavjainwiz commented:
> > You might want to take a look at #1451 (where I refactored a bit the DTOs - I've moved to a `ResourceStatusDTO`, removing the `BridgeStatusDTO` and the `ProcessorStatusDTO`)
> 
> got it.. please merge your PR first. I will merge my PR over your's.

@r00ta @manstis I had incorporated the changes introduced with #1451. Could you please review.

r00ta commented on shard-operator-parent/shard-operator-v2/src/main/java/com/redhat/service/smartevents/shard/operator/v2/ManagedProcessorSyncServiceImpl.java:
```diff
@@ -0,0 +1,96 @@
+    ManagedProcessorService managedProcessorService;
+
+    @Override
+    public void syncManagedProcessorWithManager() {
+        managerClient.fetchProcessorsForDataPlane()
+                .onItem()
+                .invoke(this::processDelta)
+                .subscribe().with(
+                        success -> LOGGER.debug("Successfully processed ManagedProcessor deltas"),
+                        error -> LOGGER.error("Failed to process ManagedProcessor deltas"));
+    }
+
+    @Override
+    public void syncManagedProcessorStatusBackToManager() {
+        managerClient.fetchProcessorsForDataPlane()
+                .onItem()
+                .transform(this::transformToProcessorStatus)
+                .invoke(this::notifyProcessorStatus)
+                .subscribe().with(
+                        success -> LOGGER.debug("Successfully sync ManagedProcessor status with Manager"),
+                        error -> LOGGER.error("Failed to sync ManagedProcessor status with Manager", error));
+    }
```
Might be a good idea to avoid calling `managerClient.fetchProcessorsForDataPlane()` when you want to sync back the status as this would force the manager to fetch all the instances from the DB another time - it is not really necessary

vaibhavjainwiz replied:
We already have a discussion about it and its finalize to keep both processes seperate.

r00ta replied:
Thanks, could you point me to the motivations/discussions or provide a good reason here to do it?

r00ta commented on shard-operator-parent/shard-operator-v2/src/main/java/com/redhat/service/smartevents/shard/operator/v2/ManagerSyncService.java:
```diff
@@ -30,5 +30,6 @@ public void syncUpdatesFromManager() {
     public void syncStatusBackToManager() {
         LOGGER.debug("Sending back status from Operator to Manager");
         managedBridgeSyncService.syncManagedBridgeStatusBackToManager();
+        managedProcessorSyncService.syncManagedProcessorStatusBackToManager();
```
Related to the previous comment: you could fetch all the instances from the manager, then with `onItem()` process all the instances and with another `onItem()` sync the status back to the manager. Just saying, you can do it with a single chain of `Uni`s without fetching all the instances again from the manager

vaibhavjainwiz replied:
please refer to my comment above

manstis commented on shard-operator-parent/shard-operator-core/src/test/java/com/redhat/service/smartevents/shard/operator/core/resources/CustomResourceStatusTest.java:
```diff
@@ -52,25 +44,6 @@ public void testMarkConditionFailure() {
     }
 
-    @ParameterizedTest
-    @MethodSource("inferManagedResourceStatusParams")
```
You can also delete the "method source" [here](https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/blob/c73169cc6cb7b44e0405ff1cc70347a53cce9217/shard-operator-parent/shard-operator-core/src/test/java/com/redhat/service/smartevents/shard/operator/core/resources/CustomResourceStatusTest.java#L47-L53).

vaibhavjainwiz replied:
thanks for pointed it out. I removed it.

manstis commented on shard-operator-parent/shard-operator-v2/src/main/java/com/redhat/service/smartevents/shard/operator/v2/ManagedProcessorServiceImpl.java:
```diff
@@ -1,19 +1,59 @@
     @Override
     public void createManagedProcessor(ProcessorDTO processorDTO) {
-        // @TODO
+        String expectedNamespace = namespaceProvider.getNamespaceName(processorDTO.getBridgeId());
```
This looks to relate to https://issues.redhat.com/browse/MGDOBR-1251 and not https://issues.redhat.com/browse/MGDOBR-1252 (which is the subject of this PR). Have you chosen to merge the work into a single PR? If so can you comment the JIRA or something so it is clear.

vaibhavjainwiz replied:
Implementation of [MGDOBR-1251](https://issues.redhat.com/browse/MGDOBR-1251) was done my mocking service layer. I had added the implementation of these methods in [MGDOBR-1252](https://issues.redhat.com/browse/MGDOBR-1252) to incorporate Rob's review comment.

manstis commented on shard-operator-parent/shard-operator-v2/src/test/java/com/redhat/service/smartevents/shard/operator/v2/ManagedProcessorServiceTest.java:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    @Test
+    public void TestCreateManagedProcessor() {
```
If I was being picky I'd say the method name should be camel-case not pascal-case.

manstis commented on shard-operator-parent/shard-operator-v2/src/test/java/com/redhat/service/smartevents/shard/operator/v2/ManagedProcessorServiceTest.java:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    @Test
+    public void TestDeleteManagedProcessor() {
```
If I was being picky I'd say the method name should be camel-case not pascal-case.

manstis commented on shard-operator-parent/shard-operator-v2/src/test/java/com/redhat/service/smartevents/shard/operator/v2/ManagedProcessorServiceTest.java:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    @Test
+    public void TestFetchAllManagedProcessors() {
```
If I was being picky I'd say the method name should be camel-case not pascal-case.

vaibhavjainwiz replied:
done

manstis commented on shard-operator-parent/shard-operator-v2/src/test/java/com/redhat/service/smartevents/shard/operator/v2/ManagedProcessorSyncServiceTest.java:
```diff
@@ -0,0 +1,113 @@
+    }
+
+    @Test
+    public void TestSyncManagedProcessorStatusBackToManager() {
```
If I was being picky I'd say the method name should be camel-case not pascal-case.

vaibhavjainwiz replied:
done

manstis commented on shard-operator-parent/shard-operator-v2/src/test/java/com/redhat/service/smartevents/shard/operator/v2/ManagerClientTest.java:
```diff
@@ -92,4 +89,26 @@ public void TestToNotifyBridgeStatus() {
     }
+
+    @Test
+    public void TestToNotifyProcessorStatus() {
```
If I was being picky I'd say the method name should be camel-case not pascal-case.

vaibhavjainwiz replied:
done

manstis reviewed (commented):
A handful of minor comments.

manstis reviewed (approved):
LGTM :+1:

manstis commented:
@vaibhavjainwiz I think it is customary to get at least one QE to approve PRs too before merge.

I know you had "two (engineering) approvals" and as such _technically_ that's enough.. but QE add a lot of value.

QE are expanding their involvement (see email "Transitioning from QE to a new Quality Community").

IMO should always be included even if it delays merge.
